File: src/clients/go/pkg/circuitbreaker/circuitbreaker_test.go
// Package circuitbreaker tests per client-retry/spec.md requirements.
package circuitbreaker

import (
	"sync"
	"testing"
	"time"
)

func TestDefaultConfig(t *testing.T) {
	// Per spec: 50% failure rate, 10 min requests, 10s window, 30s open, 5 half-open requests
	config := DefaultConfig()

	if config.FailureThreshold != 0.5 {
		t.Errorf("FailureThreshold = %v, want 0.5", config.FailureThreshold)
	}
	if config.MinRequests != 10 {
		t.Errorf("MinRequests = %v, want 10", config.MinRequests)
	}
	if config.WindowDurationMs != 10000 {
		t.Errorf("WindowDurationMs = %v, want 10000", config.WindowDurationMs)
	}
	if config.OpenDurationMs != 30000 {
		t.Errorf("OpenDurationMs = %v, want 30000", config.OpenDurationMs)
	}
	if config.HalfOpenRequests != 5 {
		t.Errorf("HalfOpenRequests = %v, want 5", config.HalfOpenRequests)
	}
}

func TestCircuitBreakerInitialState(t *testing.T) {
	cb := New("test-replica")

	if cb.State() != Closed {
		t.Errorf("Initial state = %v, want Closed", cb.State())
	}
	if !cb.IsClosed() {
		t.Error("IsClosed() = false, want true")
	}
	if cb.IsOpen() {
		t.Error("IsOpen() = true, want false")
	}
	if !cb.AllowRequest() {
		t.Error("AllowRequest() = false, want true (circuit closed)")
	}
}

func TestCircuitBreakerOpensOn50PercentFailureWith10Requests(t *testing.T) {
	// Per spec: Opens when 50% failure rate in 10s window AND >= 10 requests
	cb := New("test-replica")

	// 5 successes
	for i := 0; i < 5; i++ {
		cb.AllowRequest()
		cb.RecordSuccess()
	}

	if cb.State() != Closed {
		t.Errorf("State after 5 successes = %v, want Closed", cb.State())
	}

	// 5 failures (total 10 requests, 50% failure rate)
	for i := 0; i < 5; i++ {
		cb.AllowRequest()
		cb.RecordFailure()
	}

	if cb.State() != Open {
		t.Errorf("State after 50%% failure rate = %v, want Open", cb.State())
	}
	if !cb.IsOpen() {
		t.Error("IsOpen() = false, want true")
	}
}

func TestCircuitBreakerDoesNotOpenBelowMinRequests(t *testing.T) {
	// Per spec: Minimum requests in window before circuit can open
	cb := New("test-replica")

	// 4 failures out of 9 requests (44% failure rate, but < 10 min requests)
	for i := 0; i < 5; i++ {
		cb.AllowRequest()
		cb.RecordSuccess()
	}
	for i := 0; i < 4; i++ {
		cb.AllowRequest()
		cb.RecordFailure()
	}

	if cb.State() != Closed {
		t.Errorf("State with < 10 requests = %v, want Closed", cb.State())
	}
}

func TestCircuitBreakerRejectsRequestsWhenOpen(t *testing.T) {
	cb := New("test-replica")
	cb.ForceOpen()

	if cb.AllowRequest() {
		t.Error("AllowRequest() = true when open, want false")
	}
}

func TestCircuitBreakerTransitionsToHalfOpen(t *testing.T) {
	// Per spec: Stays open for 30 seconds before transitioning to half-open
	config := DefaultConfig()
	config.OpenDurationMs = 50 // 50ms for testing
	cb := NewWithConfig("test-replica", config)

	cb.ForceOpen()
	if cb.State() != Open {
		t.Errorf("State after ForceOpen = %v, want Open", cb.State())
	}

	// Wait for open duration
	time.Sleep(60 * time.Millisecond)

	// State check should transition to half-open
	if cb.State() != HalfOpen {
		t.Errorf("State after open duration = %v, want HalfOpen", cb.State())
	}
	if !cb.IsHalfOpen() {
		t.Error("IsHalfOpen() = false, want true")
	}
}

func TestCircuitBreakerHalfOpenLimitsRequests(t *testing.T) {
	// Per spec: Half-open allows 5 test requests
	config := DefaultConfig()
	config.HalfOpenRequests = 3 // 3 for faster testing
	config.OpenDurationMs = 10
	cb := NewWithConfig("test-replica", config)

	cb.ForceOpen()
	time.Sleep(20 * time.Millisecond)

	// First 3 requests should be allowed
	for i := 0; i < 3; i++ {
		if !cb.AllowRequest() {
			t.Errorf("AllowRequest() %d = false in half-open, want true", i)
		}
	}

	// 4th request should be rejected
	if cb.AllowRequest() {
		t.Error("AllowRequest() 4 = true in half-open, want false (limit reached)")
	}
}

func TestCircuitBreakerClosesAfterHalfOpenSuccesses(t *testing.T) {
	// Per spec: Circuit closes after sufficient successes in half-open
	config := DefaultConfig()
	config.HalfOpenRequests = 3
	config.OpenDurationMs = 10
	cb := NewWithConfig("test-replica", config)

	cb.ForceOpen()
	time.Sleep(20 * time.Millisecond)

	// All half-open requests succeed
	for i := 0; i < 3; i++ {
		cb.AllowRequest()
		cb.RecordSuccess()
	}

	if cb.State() != Closed {
		t.Errorf("State after half-open successes = %v, want Closed", cb.State())
	}
}

func TestCircuitBreakerReopensOnHalfOpenFailure(t *testing.T) {
	// Per spec: Circuit re-opens on any failure in half-open
	config := DefaultConfig()
	config.HalfOpenRequests = 5
	config.OpenDurationMs = 10
	cb := NewWithConfig("test-replica", config)

	cb.ForceOpen()
	time.Sleep(20 * time.Millisecond)

	// First request succeeds
	cb.AllowRequest()
	cb.RecordSuccess()

	// Second request fails
	cb.AllowRequest()
	cb.RecordFailure()

	if cb.State() != Open {
		t.Errorf("State after half-open failure = %v, want Open", cb.State())
	}
}

func TestCircuitBreakerError(t *testing.T) {
	err := ErrCircuitOpen{
		CircuitName:  "test-replica",
		CircuitState: Open,
	}

	msg := err.Error()
	if msg == "" {
		t.Error("Error message is empty")
	}
	if err.CircuitName != "test-replica" {
		t.Errorf("CircuitName = %v, want test-replica", err.CircuitName)
	}
	if err.CircuitState != Open {
		t.Errorf("CircuitState = %v, want Open", err.CircuitState)
	}
}

func TestCircuitBreakerMetrics(t *testing.T) {
	cb := New("test-replica")

	// Generate some activity
	for i := 0; i < 5; i++ {
		cb.AllowRequest()
		cb.RecordSuccess()
	}
	for i := 0; i < 3; i++ {
		cb.AllowRequest()
		cb.RecordFailure()
	}

	metrics := cb.Metrics()

	if metrics.State != Closed {
		t.Errorf("Metrics.State = %v, want Closed", metrics.State)
	}
	if metrics.TotalRequests != 8 {
		t.Errorf("Metrics.TotalRequests = %v, want 8", metrics.TotalRequests)
	}
	if metrics.FailedRequests != 3 {
		t.Errorf("Metrics.FailedRequests = %v, want 3", metrics.FailedRequests)
	}
}

func TestCircuitBreakerConcurrency(t *testing.T) {
	// Test thread safety
	cb := New("test-replica")
	var wg sync.WaitGroup

	// 100 concurrent operations
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func(idx int) {
			defer wg.Done()

			if cb.AllowRequest() {
				if idx%2 == 0 {
					cb.RecordSuccess()
				} else {
					cb.RecordFailure()
				}
			}

			// Read operations
			_ = cb.State()
			_ = cb.FailureRate()
			_ = cb.Metrics()
		}(i)
	}

	wg.Wait()

	// Should not panic or deadlock
	_ = cb.String()
}

func TestCircuitBreakerForceMethods(t *testing.T) {
	cb := New("test-replica")

	cb.ForceOpen()
	if cb.State() != Open {
		t.Errorf("State after ForceOpen = %v, want Open", cb.State())
	}

	cb.ForceClosed()
	if cb.State() != Closed {
		t.Errorf("State after ForceClosed = %v, want Closed", cb.State())
	}
}

func TestCircuitBreakerString(t *testing.T) {
	cb := New("test-replica")

	str := cb.String()
	if str == "" {
		t.Error("String() returned empty")
	}
}

func TestStateString(t *testing.T) {
	tests := []struct {
		state State
		want  string
	}{
		{Closed, "CLOSED"},
		{Open, "OPEN"},
		{HalfOpen, "HALF_OPEN"},
		{State(99), "UNKNOWN"},
	}

	for _, tt := range tests {
		if got := tt.state.String(); got != tt.want {
			t.Errorf("State(%d).String() = %v, want %v", tt.state, got, tt.want)
		}
	}
}

func TestCircuitBreakerFailureRate(t *testing.T) {
	cb := New("test-replica")

	// No requests
	if cb.FailureRate() != 0.0 {
		t.Errorf("FailureRate with no requests = %v, want 0.0", cb.FailureRate())
	}

	// 50% failure rate
	for i := 0; i < 5; i++ {
		cb.AllowRequest()
		cb.RecordSuccess()
	}
	for i := 0; i < 5; i++ {
		cb.AllowRequest()
		cb.RecordFailure()
	}

	rate := cb.FailureRate()
	if rate < 0.49 || rate > 0.51 {
		t.Errorf("FailureRate = %v, want ~0.5", rate)
	}
}

File: src/clients/java/src/test/java/com/archerdb/geo/CircuitBreakerTest.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2025 ArcherDB Contributors
package com.archerdb.geo;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;

/**
 * Tests for CircuitBreaker.
 *
 * Per client-retry/spec.md circuit breaker requirements: - States: Closed (normal), Open
 * (fail-fast), Half-Open (testing recovery) - Opens when: 50% failure rate in 10s window AND >= 10
 * requests - Stays open for 30 seconds before transitioning to half-open - Half-open allows 5 test
 * requests before deciding to close or re-open - Per-replica scope (not global)
 */
class CircuitBreakerTest {

    private CircuitBreaker breaker;

    @BeforeEach
    void setUp() {
        // Use short durations for testing
        breaker = CircuitBreaker.builder().setName("test-replica").setFailureThreshold(0.5)
                .setMinimumRequests(10).setWindowMs(10_000).setOpenDurationMs(100) // Short for
                                                                                   // testing
                .setHalfOpenRequests(5).build();
    }

    @Test
    @DisplayName("Initial state is CLOSED")
    void testInitialStateClosed() {
        assertEquals(CircuitBreaker.State.CLOSED, breaker.getState());
        assertTrue(breaker.isClosed());
        assertFalse(breaker.isOpen());
        assertFalse(breaker.isHalfOpen());
    }

    @Test
    @DisplayName("Requests allowed when circuit is closed")
    void testRequestsAllowedWhenClosed() {
        for (int i = 0; i < 100; i++) {
            assertTrue(breaker.allowRequest());
        }
    }

    @Test
    @DisplayName("Circuit stays closed under failure threshold")
    void testStaysClosedUnderThreshold() {
        // 9 requests with 4 failures (44%) - under 50% threshold
        for (int i = 0; i < 5; i++) {
            breaker.allowRequest();
            breaker.recordSuccess();
        }
        for (int i = 0; i < 4; i++) {
            breaker.allowRequest();
            breaker.recordFailure();
        }

        assertTrue(breaker.isClosed());
        assertEquals(4.0 / 9, breaker.getCurrentFailureRate(), 0.01);
    }

    @Test
    @DisplayName("Circuit opens after failure threshold exceeded")
    void testOpensAfterThresholdExceeded() {
        // 10 requests with 6 failures (60%) - exceeds 50% threshold
        for (int i = 0; i < 4; i++) {
            breaker.allowRequest();
            breaker.recordSuccess();
        }
        for (int i = 0; i < 6; i++) {
            breaker.allowRequest();
            breaker.recordFailure();
        }

        assertTrue(breaker.isOpen());
        assertEquals(6.0 / 10, breaker.getCurrentFailureRate(), 0.01);
    }

    @Test
    @DisplayName("Circuit rejects requests when open")
    void testRejectsRequestsWhenOpen() {
        breaker.forceOpen();

        assertFalse(breaker.allowRequest());
        assertFalse(breaker.allowRequest());
        assertFalse(breaker.allowRequest());

        assertTrue(breaker.getRejectedRequests() >= 3);
    }

    @Test
    @DisplayName("Circuit transitions to half-open after open duration")
    void testTransitionsToHalfOpen() throws InterruptedException {
        breaker.forceOpen();
        assertTrue(breaker.isOpen());

        // Wait for open duration
        Thread.sleep(150);

        // Should transition to half-open on next check
        assertEquals(CircuitBreaker.State.HALF_OPEN, breaker.getState());
        assertTrue(breaker.isHalfOpen());
    }

    @Test
    @DisplayName("Successful half-open requests close the circuit")
    void testSuccessfulHalfOpenCloses() throws InterruptedException {
        breaker.forceOpen();
        Thread.sleep(150);

        // Allow and succeed 5 requests in half-open
        for (int i = 0; i < 5; i++) {
            assertTrue(breaker.allowRequest());
            breaker.recordSuccess();
        }

        assertTrue(breaker.isClosed());
    }

    @Test
    @DisplayName("Failed half-open request reopens the circuit")
    void testFailedHalfOpenReopens() throws InterruptedException {
        breaker.forceOpen();
        Thread.sleep(150);

        // First half-open request fails
        assertTrue(breaker.allowRequest());
        breaker.recordFailure();

        assertTrue(breaker.isOpen());
    }

    @Test
    @DisplayName("Half-open limits test requests")
    void testHalfOpenLimitsRequests() throws InterruptedException {
        breaker.forceOpen();
        Thread.sleep(150);

        // Allow exactly 5 requests
        for (int i = 0; i < 5; i++) {
            assertTrue(breaker.allowRequest());
        }

        // 6th request should be rejected
        assertFalse(breaker.allowRequest());
    }

    @Test
    @DisplayName("Minimum requests required before opening")
    void testMinimumRequestsRequired() {
        // 9 requests all failures (100%) - under minimum of 10
        for (int i = 0; i < 9; i++) {
            breaker.allowRequest();
            breaker.recordFailure();
        }

        assertTrue(breaker.isClosed());
        assertEquals(1.0, breaker.getCurrentFailureRate(), 0.01);

        // 10th failure opens circuit
        breaker.allowRequest();
        breaker.recordFailure();

        assertTrue(breaker.isOpen());
    }

    @Test
    @DisplayName("Force close resets state")
    void testForceClose() {
        breaker.forceOpen();
        assertTrue(breaker.isOpen());

        breaker.forceClose();

        assertTrue(breaker.isClosed());
        assertEquals(0.0, breaker.getCurrentFailureRate(), 0.01);
    }

    @Test
    @DisplayName("State changes are tracked")
    void testStateChangesTracked() {
        assertEquals(0, breaker.getStateChanges());

        breaker.forceOpen();
        assertEquals(1, breaker.getStateChanges());

        breaker.forceClose();
        assertEquals(2, breaker.getStateChanges());
    }

    @Test
    @DisplayName("Builder validates parameters")
    void testBuilderValidation() {
        assertThrows(IllegalArgumentException.class, () -> CircuitBreaker.builder().setName(null));

        assertThrows(IllegalArgumentException.class, () -> CircuitBreaker.builder().setName(""));

        assertThrows(IllegalArgumentException.class,
                () -> CircuitBreaker.builder().setFailureThreshold(-0.1));

        assertThrows(IllegalArgumentException.class,
                () -> CircuitBreaker.builder().setFailureThreshold(1.1));

        assertThrows(IllegalArgumentException.class,
                () -> CircuitBreaker.builder().setMinimumRequests(0));

        assertThrows(IllegalArgumentException.class, () -> CircuitBreaker.builder().setWindowMs(0));

        assertThrows(IllegalArgumentException.class,
                () -> CircuitBreaker.builder().setOpenDurationMs(-1));

        assertThrows(IllegalArgumentException.class,
                () -> CircuitBreaker.builder().setHalfOpenRequests(0));
    }

    @Test
    @DisplayName("Default constants match spec requirements")
    void testDefaultConstants() {
        assertEquals(0.5, CircuitBreaker.DEFAULT_FAILURE_THRESHOLD);
        assertEquals(10, CircuitBreaker.DEFAULT_MINIMUM_REQUESTS);
        assertEquals(10_000, CircuitBreaker.DEFAULT_WINDOW_MS);
        assertEquals(30_000, CircuitBreaker.DEFAULT_OPEN_DURATION_MS);
        assertEquals(5, CircuitBreaker.DEFAULT_HALF_OPEN_REQUESTS);
    }

    @Test
    @DisplayName("toString returns meaningful string")
    void testToString() {
        String str = breaker.toString();

        assertTrue(str.contains("test-replica"));
        assertTrue(str.contains("CLOSED"));
    }

    @Test
    @DisplayName("Circuit breaker is per-replica")
    void testPerReplicaScope() {
        CircuitBreaker breaker1 = new CircuitBreaker("replica-1");
        CircuitBreaker breaker2 = new CircuitBreaker("replica-2");

        // Open breaker1
        breaker1.forceOpen();

        // breaker2 should still be closed
        assertTrue(breaker1.isOpen());
        assertTrue(breaker2.isClosed());
        assertTrue(breaker2.allowRequest());
    }

    @Test
    @DisplayName("Concurrent access is safe")
    void testConcurrentAccess() throws InterruptedException {
        int threads = 10;
        int iterations = 1000;
        Thread[] threadArray = new Thread[threads];

        for (int i = 0; i < threads; i++) {
            final int threadId = i;
            threadArray[i] = new Thread(() -> {
                for (int j = 0; j < iterations; j++) {
                    if (breaker.allowRequest()) {
                        if (threadId % 2 == 0) {
                            breaker.recordSuccess();
                        } else {
                            breaker.recordFailure();
                        }
                    }
                }
            });
        }

        for (Thread t : threadArray) {
            t.start();
        }
        for (Thread t : threadArray) {
            t.join();
        }

        // No exceptions = success
        // State should be deterministic
        assertNotNull(breaker.getState());
    }

    @Test
    @DisplayName("Failure rate at exactly threshold opens circuit")
    void testExactThresholdOpens() {
        // Exactly 50% failure rate
        for (int i = 0; i < 5; i++) {
            breaker.allowRequest();
            breaker.recordSuccess();
        }
        for (int i = 0; i < 5; i++) {
            breaker.allowRequest();
            breaker.recordFailure();
        }

        assertTrue(breaker.isOpen());
    }

    @Test
    @DisplayName("CircuitBreakerException contains correct info")
    void testCircuitBreakerException() {
        CircuitBreakerException ex =
                new CircuitBreakerException("test-circuit", CircuitBreaker.State.OPEN);

        assertEquals("test-circuit", ex.getCircuitName());
        assertEquals(CircuitBreaker.State.OPEN, ex.getCircuitState());
        assertEquals(CircuitBreakerException.CIRCUIT_OPEN, ex.getErrorCode());
        assertTrue(ex.isRetryable());
        assertTrue(ex.getMessage().contains("test-circuit"));
        assertTrue(ex.getMessage().contains("OPEN"));
    }

    @Test
    @DisplayName("Recording success in closed state records in window")
    void testSuccessRecordsInWindow() {
        for (int i = 0; i < 5; i++) {
            breaker.allowRequest();
            breaker.recordSuccess();
        }

        assertEquals(0.0, breaker.getCurrentFailureRate(), 0.01);
    }

    @Test
    @DisplayName("Mixed success/failure maintains accurate rate")
    void testMixedSuccessFailure() {
        // 3 successes, 2 failures = 40% failure rate
        for (int i = 0; i < 3; i++) {
            breaker.allowRequest();
            breaker.recordSuccess();
        }
        for (int i = 0; i < 2; i++) {
            breaker.allowRequest();
            breaker.recordFailure();
        }

        assertEquals(0.4, breaker.getCurrentFailureRate(), 0.01);
        assertTrue(breaker.isClosed()); // Under threshold
    }

    @Test
    @DisplayName("getName returns circuit name")
    void testGetName() {
        assertEquals("test-replica", breaker.getName());
    }

    @Test
    @DisplayName("Rejected requests counter increments")
    void testRejectedRequestsCounter() {
        breaker.forceOpen();

        long before = breaker.getRejectedRequests();
        breaker.allowRequest();
        breaker.allowRequest();
        long after = breaker.getRejectedRequests();

        assertEquals(before + 2, after);
    }
}
